askama_shared/src/generator.rs

```diff
@@ -1523,51 +1567,73 @@ impl Buffer {
     }
 }
 
+// type SetChain<'a, T> = MapChain<'a, T, ()>;
+
 #[derive(Debug)]
-struct SetChain<'a, T: 'a>
+struct MapChain<'a, K: 'a, V: 'a>
 where
-    T: cmp::Eq + hash::Hash,
+    K: cmp::Eq + hash::Hash,
 {
-    parent: Option<&'a SetChain<'a, T>>,
-    scopes: Vec<HashSet<T>>,
+    parent: Option<&'a MapChain<'a, K, V>>,
+    scopes: Vec<HashMap<K, V>>,
 }
 
-impl<'a, T: 'a> SetChain<'a, T>
+impl<'a, K: 'a, V: 'a> MapChain<'a, K, V>
 where
-    T: cmp::Eq + hash::Hash,
+    K: cmp::Eq + hash::Hash,
 {
-    fn new() -> SetChain<'a, T> {
-        SetChain {
+    fn new() -> MapChain<'a, K, V> {
+        MapChain {
             parent: None,
-            scopes: vec![HashSet::new()],
+            scopes: vec![HashMap::new()],
         }
     }
 
-    fn with_parent<'p>(parent: &'p SetChain<T>) -> SetChain<'p, T> {
-        SetChain {
+    fn with_parent<'p>(parent: &'p MapChain<K, V>) -> MapChain<'p, K, V> {
+        MapChain {
             parent: Some(parent),
-            scopes: vec![HashSet::new()],
+            scopes: vec![HashMap::new()],
         }
     }
 
-    fn contains(&self, val: T) -> bool {
-        self.scopes.iter().rev().any(|set| set.contains(&val))
+    fn contains(&self, key: &K) -> bool {
+        self.scopes.iter().rev().any(|set| set.contains_key(&key))
             || match self.parent {
-                Some(set) => set.contains(val),
+                Some(set) => set.contains(key),
                 None => false,
             }
     }
 
+    fn get(&self, key: &K) -> Option<&V> {
+        self.get_skip(key, 0)
+    }
+
+    /// Returns `Some` if `key` exists, while `skip` represents
+    /// the amount of previous scopes to skip in reverse.
+    fn get_skip(&self, key: &K, skip: usize) -> Option<&V> {
+        self.scopes
+            .iter()
+            .rev()
+            .skip(skip)
+            .filter_map(|set| set.get(&key))
+            .next()
+            .or_else(|| match self.parent {
+                Some(set) => set.get(key),
+                None => None,
+            })
```

Comment:
This is pretty opaque to me. Maybe name some intermediate values (which might also make it more compact)?

Reply:
Do you mean something like this (1):

```rust
let scores = self.scopes.iter().rev().skip(skip);
scores
    .filter_map(|set| set.get(&key))
    .next()
    .or_else(|| self.parent.and_then(|set| set.get(key)))
```

or (2):

```rust
let scores = self.scopes.iter().rev().skip(skip);
let var = scores.filter_map(|set| set.get(&key)).next();
var.or_else(|| self.parent.and_then(|set| set.get(key)))
```

It's more compact (less lines), but personally I find it harder to read. Unless, that wasn't what you meant? :)

Reply:
Do you find both of them harder to read, or just the latter?  I think I prefer option (1) (but I'm still a bit iffy on what this is actually intended to achieve).

Reply:
Wait, option (1) is fine, yeah.

Originally, I stored variable names in the `Option<String>` and not the resolved expression. So previously, if you did `{% call foo(a) %}` and then the macro parameter was also called `a`. Then to be able to resolve it, it required being able to `get()` while "skipping" the last `N` scopes (to be able to escape the macro's scope).

However, now that it stored the "resolved expression", then I'm feeling it might actually be possible to remove `get_skip` entirely, and then just use `get()`. I'll look into it.